### DISCUSSION TITLE
Calculation of graph property dynamically

### DISCUSSION THREAD

**Discussion by VincentLapointe-OPALRT** [networkx/networkx#5953] [Q&A] [answered]
Hello,
I'm curious too know if there are addon to networkx allowing to dynamically update some of the properties of a graph when modifying it. As a very trivial case, i would like to update the incidence matrix everytime an (or multiple) edge is added/removed. 

Moreover, I would like the keep updated several other property of the graph: cut-sets, loop-sets, BFS, spanning tree, components (disjoint-sets), proper-tree, ... 

For simple properties, it is rather easy to do such type of algorithm by updating matrixes on the fly, but as the number of properties increases I feel it is a mess to do so from scratch considering that many property are dependent of each other. And I also have a feeling that it could be really faster to just update matrix than recalculate them everything from zero.

So I would appreciate if such functionalities exist somewhere ?

**Comment by MridulS** [comment, marked as answer]
One way of doing this is could be by creating a sub class of `(Di)Graph`, overwrite the `add/remove_edge` method to update the properties (and attach them to the object). Something like:

``` python
In [1]: import networkx as nx

In [2]: class DynamicGraph(nx.Graph):
    ...:     def __init__(self, incoming_graph_data=None, **attr):
    ...:         super().__init__(incoming_graph_data, **attr)
    ...:         self.dynamic_properties = {}
    ...:     def add_edge(self, u_of_edge, v_of_edge, **attr):
    ...:         super().add_edge(u_of_edge, v_of_edge, **attr)
    ...:         self.compute()
    ...:     def compute(self):
    ...:         self.dynamic_properties["incidence_matrix"] = nx.incidence_matrix(self)
    ...:


```

NOTE: This is not tested at all! You would need to take care of all functions that add/remove nodes or edges in the graph and you could probably add some more logic to do some cacheing of the properties you want to track.

**Reply by VincentLapointe-OPALRT** [reply]
Hi MridulS,
Tx for your fast answer. Your proposal is exactly what I was considering to do. Note that in the simple case of adding an edge to a graph, we do not necessary want to recreate the incidence matrix from scratch as you propose in the last line of your code. In that simple case, you just need to add one column to the matrix and sets the two elements of that column to +/- 1. And you don't need to recalculate the  other columns since it is already set correctly. (Or you may need to permute columns if edge order changes.)

Of course, depending of the property and the algorithm in used, and also the number of changes done to the graph, It will probably be faster to recalculate the complete matrix. But I would love to find such toolbox somewhere... So I do not need to develop it...

Anyway, it would be a very nice feature for networkx. Something like : keep in sync a set of graph properties according to the change done to the sets of edges/nodes in the graph.

Let me know, if it makes sense or not?

**Reply by rossbar** [reply]
It's certainly possible, but whether or not it makes sense depends on your desired application/use-case. Generally, I don't think it makes sense to upstream this design pattern into NetworkX itself. The NetworkX graph data structures are designed to be flexible (hence the nested-dicts) and minimize storing state/recomputation requirements when a graph is modified dynamically.

Based on the original description, I'd tend to agree with @MridulS that you can achieve the desired behavior by creating a custom graph object, though I'd probably start with the mixin pattern:

```python
import networkx as nx

class AdjacencyMixin:
    @property
    def adj_matrix(self):
        return nx.adjacency_matrix(self)

class MyGraph(nx.Graph, AdjacencyMixin):
    pass

>>> G = MyGraph([(0, 1), (1, 2)])
>>> G.adj_matrix
<3x3 sparse array of type '<class 'numpy.int64'>'
        with 4 stored elements in Compressed Sparse Row format>
>>> G.add_node(10)
>>> G.adj_matrix
<4x4 sparse array of type '<class 'numpy.int64'>'
        with 4 stored elements in Compressed Sparse Row format>
```

The above begs the question: is extending the graph object itself *really* beneficial? Is it really that much harder to write `nx.adjacency_matrix(G)` than `G.adj_matrix` whenever you want the sparse adjacency representation? I'd argue no, but that's in the eye of the beholder :). In the case where you want to do "smarter" incremental updating (e.g. not recompute everything from scratch each time), then you're really looking at a custom graph object and you're back to the original question: what's the specific use-case? Does it make sense to try to modify/extend data structures, or can you achieve what you want with a more functional pattern?

**Reply by VincentLapointe-OPALRT** [reply]
Thank you @rossbar for your interesting perspectives. I cannot argue that it doesn't make sense for "networkx" users. 
As you mentioned, I'm really targeting a smart incremental update. The calculation of various properties is really the first stage of my math calculation. So i want to avoid rebuilding all properties for performance consideration. 

Anyway. Tx for your suggestion. I will sleep on that !
